detect/detect.go

```diff
@@ -20,38 +20,32 @@ var (
 
 // FromFile takes a filepath & tries to work out the corresponding dataset
 // for the sake of speed, it only works with files that have a recognized extension
-func FromFile(path string) (ds *dataset.Structure, err error) {
+func FromFile(path string) (st *dataset.Structure, err error) {
 	f, err := os.Open(path)
 	if err != nil {
 		return nil, err
 	}
-	return FromReader(path, f)
-}
 
-// FromReader is a shorthand for a path/filename and reader
-func FromReader(path string, data io.Reader) (ds *dataset.Structure, err error) {
 	format, err := ExtensionDataFormat(path)
 	if err != nil {
 		return nil, err
 	}
-	return Structure(format, data)
+
+	defer f.Close()
```

Comment:
Move this closer to os.Open, after its err is checked. As it is now, if ExtensionDataFormat returns an error, the file is not Closed.